flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/JobExceptionsHandlerTest.java

```diff
@@ -37,68 +36,258 @@
+                        taskManagerLocationToString(otherFailure.getAssignedResourceLocation())));
+        assertThat(response, hasHistoryEntryCount(2));
+        assertThat(response, not(hasTruncatedHistory()));
+        assertThat(response, hasTotal(2));
+    }
+
+    @Test
+    public void testWithExceptionHistoryWithTruncationThroughParameter()
+            throws HandlerRequestException {
+        final ExceptionHistoryEntry rootCause =
+                ExceptionHistoryEntry.fromGlobalFailure(
+                        new RuntimeException("exception #0"), System.currentTimeMillis());
+        final ExceptionHistoryEntry otherFailure =
+                new ExceptionHistoryEntry(
+                        new RuntimeException("exception #1"),
+                        System.currentTimeMillis(),
+                        "task name",
+                        new LocalTaskManagerLocation());
+
+        final ExecutionGraphInfo executionGraphInfo =
+                createExecutionGraphInfo(rootCause, otherFailure);
+        final HandlerRequest<EmptyRequestBody, JobExceptionsMessageParameters> request =
+                createRequest(executionGraphInfo.getJobId(), 1);
+        final JobExceptionsInfoWithHistory response =
+                testInstance.handleRequest(request, executionGraphInfo);
+
+        assertThat(response, containsRootCause(rootCause.getException(), rootCause.getTimestamp()));
+        assertThat(
+                response,
+                containsGlobalFailure(rootCause.getException(), rootCause.getTimestamp()));
+        assertThat(response, hasHistoryEntryCount(1));
+        assertThat(response, hasTruncatedHistory());
+        assertThat(response, hasTotal(2));
+    }
+
+    @Test
+    public void testWithExceptionHistoryWithInternalTruncation() throws HandlerRequestException {
+        final ExceptionHistoryEntry rootCause =
+                ExceptionHistoryEntry.fromGlobalFailure(
+                        new RuntimeException("exception #0"), System.currentTimeMillis());
+
+        final ExecutionGraphInfo executionGraphInfo = createExecutionGraphInfo(rootCause);
+        // adding the rootCause a second time here will result in the underlying BoundedFIFOQueue to
+        // be truncated
+        executionGraphInfo.getExceptionHistory().add(rootCause);
+        final HandlerRequest<EmptyRequestBody, JobExceptionsMessageParameters> request =
+                createRequest(executionGraphInfo.getJobId(), 10);
+        final JobExceptionsInfoWithHistory response =
+                testInstance.handleRequest(request, executionGraphInfo);
+
+        assertThat(response, containsRootCause(rootCause.getException(), rootCause.getTimestamp()));
+        assertThat(
+                response,
+                containsGlobalFailure(rootCause.getException(), rootCause.getTimestamp()));
+        assertThat(response, hasHistoryEntryCount(1));
+        assertThat(response, not(hasTruncatedHistory()));
```

Comment:
I see what you're saying. The `truncated` value is solely used by frontend to determine whether it can get even more entries by increasing the `GET` parameter `maxExceptions`. One way to resolve that would be to rename the `truncate` flag into something more meaningful. This comes with the problem that we would have to do a larger refactoring since we don't want to break the existing REST response schema.

But I'm gonna leave the resolution of this issue for now as it would not be an issue if [we decide that the total exception count is not a feature we want](https://github.com/apache/flink/pull/15049#discussion_r590203591).

Reply:
This test was removed as part of the `elementCount` revert